### ISSUE TITLE
[YSQL] Read restart error encountered during cache refresh might be treated as a dropped database

### ISSUE THREAD

**Issue by frozenspider** [yugabyte/yugabyte-db#15790]
Jira Link: [DB-5153](https://yugabyte.atlassian.net/browse/DB-5153)
### Description

When cache is reset/refreshed due to an outdated catalog version, the following call stack is hit:
```
YBIsDBConnectionValid + 1468
YBPreloadRelCache + 192
YbPreloadRelCacheHelper + 140
YBRefreshCache + 456
YBCheckSharedCatalogCacheVersion + 316
PostgresMain + 2716
report_fork_failure_to_client + 0
BackendStartup + 668
ServerLoop + 1016
PostmasterMain + 6808
startup_hacks + 0
main + 36
start + 520
```
This `YBIsDBConnectionValid` is supposed to treat `Not Found` errors as database missing, but it doesn't differentiate between error types, so it this call hits `Restart read required` error, client sees
```
FATAL: Could not reconnect to database
  Hint: Database might have been dropped by another user
```
`org.yb.pgsql.TestYsqlUpgrade#upgradeIsIdempotentSingleConn` is a flaky test that highlights this problem, since it has about ~30% failute rate on my machine it's better to run it repeatedly:
```
./yb_build.sh --java-test 'org.yb.pgsql.TestYsqlUpgrade#upgradeIsIdempotentSingleConn' --tp 1 -n 10 -- --skip-log-compression --stop-at-failure
```

[DB-5153]: https://yugabyte.atlassian.net/browse/DB-5153?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by frozenspider** [comment]
Underlying error will be investigated in #14519

**Comment by d-uspenskiy** [comment]
Fixed in context of #14519
